Extend/Controller/OrderController.php

```diff
@@ -238,16 +239,34 @@ private function _handleResponse($oResponse, $oLogger, $oOrder)
         if ($oResponse instanceof FailureResponse) {
             $oLogger->error('Error processing transaction:');
 
+            $aErrorDescriptions = [];
+
             foreach ($oResponse->getStatusCollection() as $oStatus) {
                 /**
                  * @var Status $oStatus
                  */
                 $sSeverity = ucfirst($oStatus->getSeverity());
                 $sCode = $oStatus->getCode();
                 $sDescription = $oStatus->getDescription();
+                $aErrorDescriptions[] = $sDescription;
                 $oLogger->error("\t$sSeverity with code $sCode and message '$sDescription' occurred.");
             }
-            return;
+
+            // set the custom payment error code and text and redirect back to the payment step of the checkout process
+            Registry::getSession()->setVariable('payerror', '-102');
+            Registry::getSession()->setVariable('payerrortext', join('<br/>', $aErrorDescriptions));
+
+            $sBaseLanguage = 'lang= ' . Registry::getLang()->getBaseLanguage();
+            $sSid = Registry::getSession()->sid(true);
+
+            $aQueryParams = [
+                'cl' => 'payment',
+                'lang' => $sBaseLanguage,
+            ];
+
+            $sFullQueryString = http_build_query($aQueryParams) . '&' . $sSid;
+
+            return Registry::getUtils()->redirect(Helper::getShopUrl() . 'index.php?' . $sFullQueryString);
```

Comment:
There is getShopHomeUrl() method on Config. Couldn't it be reused?

Reply:
Thanks for the tip, this reduced the code significantly :)